node/src/engine/engine.rs

```diff
@@ -60,47 +67,106 @@ impl TonNodeEngine {
             message_queue.clone(),
             documents_db.clone(),
             &*storage,
+            debug_mode,
         )?;
-        Ok(TonNodeEngine {
+        if workchain.finality_was_loaded {
+            masterchain.restore_state()?;
+        }
+        let node = TonNodeEngine {
             time: RwLock::new(BlockTime::new()),
             message_queue: message_queue.clone(),
             masterchain,
             workchain,
-        })
+            is_running_in_background: AtomicBool::new(false),
+        };
+        node.execute_queued_messages()?;
+        Ok(node)
     }
 
     pub fn start(self: Arc<Self>) -> NodeResult<()> {
-        if self.workchain.finality_was_loaded {
-            self.masterchain.restore_state()?;
-        }
+        self.is_running_in_background.store(true, Ordering::Relaxed);
         thread::spawn(move || loop {
-            if let Err(err) = self.process_messages(true) {
+            if let Err(err) = self.execute_queued_messages() {
                 log::error!(target: "node", "failed block generation: {}", err);
             }
-            self.message_queue.wait_new_message();
+            if self.message_queues_are_empty() {
+                self.message_queue.wait_new_message();
+            }
         });
         Ok(())
     }
 
-    pub(crate) fn process_messages(&self, debug: bool) -> NodeResult<()> {
+    fn message_queues_are_empty(&self) -> bool {
+        self.workchain.out_message_queue_is_empty()
+            && self.masterchain.out_message_queue_is_empty()
+            && self.message_queue.is_empty()
+    }
+
+    fn execute_queued_messages(&self) -> NodeResult<()> {
         let mut continue_processing = true;
         let mut gen_time = self.get_next_time();
         while continue_processing {
             continue_processing = false;
-            while let Some(block) = self.workchain.generate_block(gen_time, debug)? {
+            while let Some(block) = self.workchain.generate_block(gen_time, self.time_mode())? {
                 self.set_last_time(gen_time);
                 self.masterchain.register_new_shard_block(&block)?;
                 continue_processing = true;
                 gen_time = self.get_next_time();
             }
-            if self.masterchain.generate_block(gen_time, debug)?.is_some() {
+            if self
+                .masterchain
+                .generate_block(gen_time, self.time_mode())?
+                .is_some()
+            {
                 self.set_last_time(gen_time);
                 continue_processing = true;
             }
         }
         Ok(())
     }
 
+    pub fn process_messages(&self) -> NodeResult<()> {
```

Comment:
As I understand this functions can be called by external crates and not used in evernode SE operating. Better separate them into another `impl` section and write comment to point it out